ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/SignatureVerificationService.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2021 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.statetransition.validation;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import org.apache.tuweni.bytes.Bytes;
+import tech.pegasys.teku.bls.BLS;
+import tech.pegasys.teku.bls.BLSPublicKey;
+import tech.pegasys.teku.bls.BLSSignature;
+import tech.pegasys.teku.bls.BLSSignatureVerifier;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.service.serviceutils.Service;
+import tech.pegasys.teku.service.serviceutils.ServiceCapacityExceededException;
+import tech.pegasys.teku.spec.logic.common.util.AsyncBLSSignatureVerifier;
+
+public class SignatureVerificationService extends Service implements AsyncBLSSignatureVerifier {
+  private static final int QUEUE_CAPACITY = 1000;
+  private static final int MAX_BATCH_SIZE = 50;
+  private static final Duration INACTIVITY_PAUSE = Duration.ofMillis(200);
+
+  private final ExecutorFactory executorFactory;
+  private final int numThreads;
+  private final int maxBatchSize;
+
+  private final BlockingQueue<SignatureTask> batchSignatureTasks;
+  private ScheduledExecutorService executor;
+
+  @VisibleForTesting
+  SignatureVerificationService(
+      final ExecutorFactory executorFactory,
+      final int numThreads,
+      final int queueCapacity,
+      final int maxBatchSize) {
+    this.executorFactory = executorFactory;
+    this.numThreads = Math.min(numThreads, Runtime.getRuntime().availableProcessors());
+    this.maxBatchSize = maxBatchSize;
+
+    this.batchSignatureTasks = new ArrayBlockingQueue<SignatureTask>(queueCapacity);
+  }
+
+  public static SignatureVerificationService create() {
+    return new SignatureVerificationService(
+        SignatureVerificationService::defaultExecutorFactory, 2, QUEUE_CAPACITY, MAX_BATCH_SIZE);
+  }
+
+  private static ScheduledExecutorService defaultExecutorFactory(final int numThreads) {
+    return new ScheduledThreadPoolExecutor(
```

Comment:
I suspect this would be a good place to use a fixed thread pool executor (although doing so assumes you agree with the polling approach I suggest below so we don't need a ScheduledThreadPoolExecutor):
```
return Executors.newFixedThreadPool(
        numThreads,
        new ThreadFactoryBuilder()
            .setDaemon(false)
            .setNameFormat(SignatureVerificationService.class.getSimpleName() + "-%d")
            .build());
```

We could potentially use a dedicated `AsyncRunner` for this now as well (created from the `AsyncRunnerFactory`), so it follows the usual patterns and the factory would guarantee it gets shutdown. Don't think it really makes much difference though.